pkg/ansible/controller/reconcile.go

```diff
@@ -0,0 +1,172 @@
+		logrus.Debugf("Adding finalizer %s to resource", finalizer)
+		finalizers := append(pendingFinalizers, finalizer)
+		u.SetFinalizers(finalizers)
+		err := r.Client.Update(context.TODO(), u)
+		return reconcile.Result{}, err
+	}
+	if !contains(pendingFinalizers, finalizer) && deleted {
+		logrus.Info("Resource is terminated, skipping reconcilation")
+		return reconcile.Result{}, nil
+	}
+
+	s := u.Object["spec"]
+	_, ok := s.(map[string]interface{})
+	if !ok {
+		logrus.Warnf("spec was not found")
+		u.Object["spec"] = map[string]interface{}{}
+		r.Client.Update(context.TODO(), u)
+		return reconcile.Result{Requeue: true}, nil
+	}
+	ownerRef := metav1.OwnerReference{
+		APIVersion: u.GetAPIVersion(),
+		Kind:       u.GetKind(),
+		Name:       u.GetName(),
+		UID:        u.GetUID(),
+	}
+
+	kc, err := kubeconfig.Create(ownerRef, "http://localhost:8888", u.GetNamespace())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	defer os.Remove(kc.Name())
+	eventChan, err := r.Runner.Run(u, kc.Name())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// iterate events from ansible, looking for the final one
+	statusEvent := eventapi.StatusJobEvent{}
+	for event := range eventChan {
+		for _, eHandler := range r.EventHandlers {
+			go eHandler.Handle(u, event)
+		}
+		if event.Event == "playbook_on_stats" {
+			// convert to StatusJobEvent; would love a better way to do this
+			data, err := json.Marshal(event)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			err = json.Unmarshal(data, &statusEvent)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+	if statusEvent.Event == "" {
+		err := errors.New("did not receive playbook_on_stats event")
+		logrus.Error(err.Error())
+		return reconcile.Result{}, err
+	}
+
+	// We only want to update the CustomResource once, so we'll track changes and do it at the end
+	var needsUpdate bool
+	runSuccessful := true
+	for _, count := range statusEvent.EventData.Failures {
+		if count > 0 {
+			runSuccessful = false
+			break
+		}
+	}
+	// The finalizer has run successfully, time to remove it
+	if deleted && finalizerExists && runSuccessful {
+		finalizers := []string{}
+		for _, pendingFinalizer := range pendingFinalizers {
+			if pendingFinalizer != finalizer {
+				finalizers = append(finalizers, pendingFinalizer)
+			}
+		}
+		u.SetFinalizers(finalizers)
+		needsUpdate = true
+	}
+
+	statusMap, ok := u.Object["status"].(map[string]interface{})
+	if !ok {
+		u.Object["status"] = ResourceStatus{
+			Status: NewStatusFromStatusJobEvent(statusEvent),
+		}
+		logrus.Infof("adding status for the first time")
+		needsUpdate = true
+	} else {
+		// Need to conver the map[string]interface into a resource status.
+		if update, status := UpdateResourceStatus(statusMap, statusEvent); update {
+			u.Object["status"] = status
+			needsUpdate = true
+		}
+	}
+	if needsUpdate {
+		err = r.Client.Update(context.TODO(), u)
```

Comment:
Don't need to change this now but it might be worth looking into the Status client to see if there's a benefit to using that if we're just updating the status:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/unstructured_client.go#L136-L151
The default manager client comes with a status writer:
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/split.go#L32

Reply:
Added an issue to track this, and assigned it to me. 
https://github.com/operator-framework/operator-sdk/issues/540